storage/SMAIndexSubBlock.cpp

```diff
@@ -0,0 +1,665 @@
+ *        of the sum. Does nothing if the entry's type cannot be summed.
+ * @note We ignore clearing any old data held in the previous sum because Double
+ *       and Long types are always represented inline.
+ *       
+ * @param entry A pointer to the entry to modify.
+ */
+inline void setTypedValueForSum(SMAEntry *entry) {
+  TypeID sum_type = sumType(entry->type_id);
+  if (sum_type == kLong) {
+    new (&entry->sum_aggregate) TypedValue(static_cast<std::int64_t>(0));
+  } else if (sum_type == kDouble) {
+    new (&entry->sum_aggregate) TypedValue(static_cast<double>(0.0));
+  }
+}
+
+/**
+ * @brief Correctly zeroes the SMA entry for each possible type.
+ *
+ * @param entry The entry for which the min/max references will be zeroed out.
+ */
+void setTypedValueForMinMax(SMAEntry *entry) {
+  TypedValue *min = &(entry->min_entry_ref.value);
+  TypedValue *max = &(entry->max_entry_ref.value);
+  switch (entry->type_id) {
+    case kInt:
+      new (min) TypedValue(static_cast<int>(0));
+      new (max) TypedValue(static_cast<int>(0));
+      return;
+    case kLong:
+      new (min) TypedValue(static_cast<std::int64_t>(0));
+      new (min) TypedValue(static_cast<std::int64_t>(0));
+      return;
+    case kFloat:
+      new (min) TypedValue(static_cast<float>(0));
+      new (max) TypedValue(static_cast<float>(0));
+    case kDouble:
+      new (min) TypedValue(static_cast<double>(0.0));
+      new (max) TypedValue(static_cast<double>(0.0));
+      return;
+    case kDatetimeInterval: {
+      DatetimeIntervalLit zero;
+      zero.interval_ticks = 0;
+      new (min) TypedValue(zero);
+      new (max) TypedValue(zero);
+      return;
+    }
+    case kYearMonthInterval: {
+      YearMonthIntervalLit zero;
+      zero.months = 0;
+      new (min) TypedValue(zero);
+      new (max) TypedValue(zero);
+      return;
+    }
+    default: {
+      CHECK(false) << "SMA Index was created on an unindexable type.";
+    }
+  }
+}
+
+}  // namespace sma_internal
+
+
+SMAIndexSubBlock::SMAIndexSubBlock(const TupleStorageSubBlock &tuple_store,
+                                   const IndexSubBlockDescription &description,
+                                   const bool new_block,
+                                   void *sub_block_memory,
+                                   const std::size_t sub_block_memory_size)
+    : IndexSubBlock(tuple_store,
+                    description,
+                    new_block,
+                    sub_block_memory,
+                    sub_block_memory_size),
+      header_(nullptr),
+      entries_(nullptr),
+      total_attributes_(tuple_store.getRelation().size()),
+      attribute_to_entry_(nullptr),
+      indexed_attributes_(0),
+      initialized_(false),
+      add_operations_(new UncheckedBinaryOperator*[kNumTypeIDs]),
+      less_comparisons_(new UncheckedComparator*[kNumTypeIDs]),
+      equal_comparisons_(new UncheckedComparator*[kNumTypeIDs]) {
```

Comment:
@hbdeshmukh  makes a good point. @cramja Since we generally used STL containers, can we do that here? Lots of advantages with that approach, including not having to worry about memory management and consistency with the remaining code base.